src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1065,923 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+
+    # check that ~/.kube/config is only read- and writable by its owner
+    if platform.system() != 'Windows':
+        existing_file_perms = "{:o}".format(stat.S_IMODE(os.lstat(existing_file).st_mode))
+        if not existing_file_perms.endswith('600'):
+            logger.warning('%s has permissions "%s".\nIt should be readable and writable only by its owner.',
+                           existing_file, existing_file_perms)
+
+    with open(existing_file, 'w+') as stream:
+        try:
+            yaml.safe_dump(existing, stream, default_flow_style=False)
+        except Exception as e:
+            telemetry.set_exception(exception=e, fault_type=consts.Failed_To_Merge_Kubeconfig_File,
+                                    summary='Exception while merging the kubeconfig file')
+            raise CLIError('Exception while merging the kubeconfig file.' + str(e))
+
+    current_context = addition.get('current-context', 'UNKNOWN')
+    msg = 'Merged "{}" as current context in {}'.format(current_context, existing_file)
+    print(msg)
+
+
+def handle_merge(existing, addition, key, replace):
+    if not addition[key]:
+        return
+    if existing[key] is None:
+        existing[key] = addition[key]
+        return
+
+    for i in addition[key]:
+        for j in existing[key]:
+            if i['name'] == j['name']:
+                if replace or i == j:
+                    existing[key].remove(j)
+                else:
+                    msg = 'A different object named {} already exists in your kubeconfig file.\nOverwrite?'
+                    overwrite = False
+                    try:
+                        overwrite = prompt_y_n(msg.format(i['name']))
+                    except NoTTYException:
+                        pass
+                    if overwrite:
+                        existing[key].remove(j)
+                    else:
+                        msg = 'A different object named {} already exists in {} in your kubeconfig file.'
+                        telemetry.set_exception(exception='A different object with same name exists in the kubeconfig file', fault_type=consts.Different_Object_With_Same_Name_Fault_Type,
+                                                summary=msg.format(i['name'], key))
+                        raise CLIError(msg.format(i['name'], key))
+        existing[key].append(i)
+
+
+def _resolve_service_principal(client, identifier):  # Uses service principal graph client
+    # todo: confirm with graph team that a service principal name must be unique
+    result = list(client.list(filter="servicePrincipalNames/any(c:c eq '{}')".format(identifier)))
+    if result:
+        return result[0].object_id
+    if utils.is_guid(identifier):
+        return identifier  # assume an object id
+    error = CLIError("Service principal '{}' doesn't exist".format(identifier))
+    error.status_code = 404  # Make sure CLI returns 3
+    raise error
+
+
+def client_side_proxy_wrapper(cmd,
```

Comment:
>client_side_proxy_wrapper [](start = 4, length = 25)

Does the telemetry as to how long this command was running captured? Generally CLI framework captures duration of all commands however in this case generally user would hit ctrl+c and hence not sure.

Reply:
Also do we print what command to run to exit when user executes this command?

---
In reply to: [596740080](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r596740080) [](ancestors = 596740080)

Reply:
I'd thought of doing this as follows:
start_time = time.time()
    
def ctrlc_handler(sig, frame):
    print('command ran for {} seconds'.format(time.time()-start_time))
    telemetry.set_debug_info('Command ran for {} seconds.'.format(time.time()-start_time))
    interrupt_main()

signal.signal(signal.SIGINT, ctrlc_handler)
Issue is print statement(had kept for debug purposes) is getting executed multiple times(even more than the no of threads we are spawning) and hence telemetry would also get captured multiple times. I think we will need time to figure out the right approach for this.

Reply:
I have added a msg after proxy starts notifying the user how to close.

Reply:
What are the no of threads we are talking about here?

---
In reply to: [596830408](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r596830408) [](ancestors = 596830408)

Reply:
There are 3 threads(main thread, thread that runs csp, thread that refreshes the HC details)